File: lib/loom/board/github.ex
defmodule Loom.Board.GitHub do
  @moduledoc "Read-only GitHub CLI adapter for Board labels and item state."

  alias Loom.Board

  def adapter(root, opts \\ []) do
    runner = Keyword.get(opts, :runner, &System.cmd/3)

    %{
      claimable: &claimable(root, &1, runner),
      state: &state(root, &1, runner)
    }
  end

  defp claimable(root, :implementor, runner) do
    with {:ok, rework} <- list(root, "pr", "loom:rework", runner),
         {:ok, ready} <- list(root, "issue", "loom:ready", runner) do
      case Board.choose(:implementor, rework ++ ready) do
        nil -> :none
        item -> {:ok, item}
      end
    end
  end

  defp claimable(root, :reviewer, runner) do
    with {:ok, items} <- list(root, "pr", "loom:review", runner) do
      case Board.choose(:reviewer, items) do
        nil -> :none
        item -> {:ok, item}
      end
    end
  end

  defp state(root, item, runner) do
    kind = if item.kind == :issue, do: "issue", else: "pr"

    case runner.("gh", [kind, "view", to_string(item.id), "--json", "labels,state"],
           cd: root,
           stderr_to_stdout: true
         ) do
      {output, 0} ->
        decoded = :json.decode(output)
        {:ok, stage(decoded["labels"] || [])}

      {output, _status} ->
        {:error, {:github, String.trim(output)}}
    end
  rescue
    error -> {:error, {:decode, Exception.message(error)}}
  end

  defp list(root, kind, label, runner) do
    fields =
      if kind == "pr", do: "number,title,url,labels,headRefName", else: "number,title,url,labels"

    case runner.(
           "gh",
           [
             kind,
             "list",
             "--label",
             label,
             "--state",
             "open",
             "--json",
             fields,
             "--limit",
             "100"
           ],
           cd: root,
           stderr_to_stdout: true
         ) do
      {output, 0} ->
        items =
          output
          |> :json.decode()
          |> Enum.map(&normalize(&1, kind, label))

        {:ok, items}

      {output, _status} ->
        {:error, {:github, String.trim(output)}}
    end
  rescue
    error -> {:error, {:decode, Exception.message(error)}}
  end

  defp normalize(item, kind, label) do
    %{
      id: item["number"],
      kind: if(kind == "pr", do: :pr, else: :issue),
      title: item["title"],
      url: item["url"],
      slug: item["headRefName"] || slug(item["title"]),
      stage: label_stage(label)
    }
  end

  defp stage(labels) do
    names = Enum.map(labels, & &1["name"])

    cond do
      "loom:rework" in names -> :rework
      "loom:ready" in names -> :ready
      "loom:review" in names -> :review
      "loom:done" in names -> :done
      true -> :untracked
    end
  end

  defp label_stage("loom:rework"), do: :rework
  defp label_stage("loom:ready"), do: :ready
  defp label_stage("loom:review"), do: :review

  defp slug(title) do
    title
    |> String.downcase()
    |> String.replace(~r/[^a-z0-9]+/, "-")
    |> String.trim("-")
  end
end

File: extensions/loom-workers/github.ts
import type { BoardItem, Lifecycle, Role } from "./types.ts";

export interface CommandResult {
  stdout: string;
  stderr?: string;
  code?: number;
}

export type GhRunner = (args: string[]) => Promise<CommandResult>;

const missingObject = /not found|could not resolve to/i;

const searches: Array<{ lifecycle: Lifecycle; kind: "issue" | "pr" }> = [
  { lifecycle: "ready", kind: "issue" },
  { lifecycle: "review", kind: "pr" },
  { lifecycle: "rework", kind: "pr" },
  { lifecycle: "done", kind: "pr" },
];

export class GitHubBoard {
  private readonly repo: string;
  private readonly run: GhRunner;

  constructor(repo: string, run: GhRunner) {
    this.repo = repo;
    this.run = run;
  }

  async listOpen(): Promise<BoardItem[]> {
    const items: BoardItem[] = [];
    for (const search of searches) {
      const result = await this.run([
        search.kind, "list", "--repo", this.repo, "--state", "open",
        "--label", `loom:${search.lifecycle}`, "--limit", "10000",
        "--json", search.kind === "pr"
          ? "number,title,url,createdAt,state,labels,headRefName"
          : "number,title,url,createdAt,state,labels",
      ]);
      if ((result.code ?? 0) !== 0) throw new Error(result.stderr || `gh exited ${result.code}`);
      const rows = parseRows(result.stdout);
      for (const row of rows) {
        const labels = labelNames(row.labels);
        if (row.state !== "OPEN" || !labels.includes(`loom:${search.lifecycle}`)) continue;
        items.push({
          kind: search.kind,
          number: row.number,
          title: row.title,
          url: row.url,
          lifecycle: search.lifecycle,
          claimed: labels.includes("loom:wip"),
          createdAt: row.createdAt,
          headRefName: row.headRefName,
          open: true,
        });
      }
    }
    return items;
  }

  async observe(item: BoardItem): Promise<BoardItem | undefined> {
    let result: CommandResult;
    try {
      result = await this.run([
        item.kind, "view", String(item.number), "--repo", this.repo,
        "--json", item.kind === "pr"
          ? "number,title,url,createdAt,state,labels,headRefName"
          : "number,title,url,createdAt,state,labels",
      ]);
    } catch (error) {
      if (missingObject.test(error instanceof Error ? error.message : String(error))) return undefined;
      throw error;
    }
    if (result.code && result.code !== 0) {
      if (missingObject.test(result.stderr ?? "")) return undefined;
      throw new Error(result.stderr || `gh exited ${result.code}`);
    }
    const row = JSON.parse(result.stdout);
    const labels = labelNames(row.labels);
    const lifecycle = (["ready", "review", "rework", "done"] as const)
      .find((name) => labels.includes(`loom:${name}`)) ?? "none";
    return {
      kind: item.kind,
      number: row.number,
      title: row.title,
      url: row.url,
      createdAt: row.createdAt,
      lifecycle,
      claimed: labels.includes("loom:wip"),
      headRefName: row.headRefName,
      open: row.state === "OPEN",
    };
  }

  async next(role: Role): Promise<BoardItem | undefined> {
    const eligible = (await this.listOpen()).filter((item) =>
      !item.claimed && (role === "reviewer"
        ? item.kind === "pr" && item.lifecycle === "review"
        : (item.kind === "issue" && item.lifecycle === "ready") ||
          (item.kind === "pr" && item.lifecycle === "rework")),
    );
    eligible.sort((left, right) => {
      const lifecycle = Number(left.lifecycle !== "rework") - Number(right.lifecycle !== "rework");
      return lifecycle || left.createdAt.localeCompare(right.createdAt) || left.number - right.number;
    });
    return eligible[0];
  }
}

function parseRows(stdout: string): any[] {
  const value = JSON.parse(stdout || "[]");
  if (!Array.isArray(value)) throw new Error("gh returned invalid Board JSON");
  return value;
}

function labelNames(labels: unknown): string[] {
  if (!Array.isArray(labels)) return [];
  return labels.map((label) => typeof label === "string" ? label : (label as any)?.name).filter(Boolean);
}
